### ISSUE TITLE
is there any implemented Negative binomial distributions likelihood loss function 

### ISSUE THREAD

**Issue by kb007** [zhykoties/DeepAR#6]
Hello, thank you that i have implemented the DeepAR model for my data set. but i want to implement negative Binomial distributions likelihood loss function since it is in the paper. I am very beginner in PyTorch but somehow I have implemented your Gaussian implementation in for my dataset. Pytorch itself has a Negative Binomial Class which contains total count, logits, probs as a parameter. actually  I cant relate which of these parameters indicates the zero index, mu, sigma I have tried with my intuitions. but  my y_prediction output is showing inappropriate value. I am looking for your suggestion which might be helpful for me to solve this

**Comment by zhykoties** [comment]
Hi, you can try the following code:

class NegBinomial:

    def __init__(self):
        self.softplus_mu = nn.Softplus()
        self.softplus_alpha = nn.Softplus()

    def loss_fn(self, parameters: torch.Tensor, v_batch: torch.Tensor, labels: torch.Tensor, missing_value = 0, use_boundary = False):
        mu = self.softplus_mu(parameters[:, :, 0])
        alpha = self.softplus_alpha(parameters[:, :, 1])
        zero_index = (labels != missing_value)
        distribution = torch.distributions.negative_binomial.NegativeBinomial(total_count=1/alpha[zero_index], logits=alpha[zero_index]*mu[zero_index])
        likelihood = distribution.log_prob(labels[zero_index])
        return -torch.mean(likelihood)

    def emit_mean(self, parameters: torch.Tensor, v_batch: torch.Tensor, use_boundary = False):
        mu = self.softplus_mu(parameters[:, :, 0])
        alpha = self.softplus_alpha(parameters[:, :, 1])
        logits = mu * alpha
        mean = 1 / alpha * torch.exp(logits)
        std = mean / torch.sigmoid(-logits)
        return mean, std

    def emit_mean_one_syep(self, parameters: torch.Tensor, v_batch: torch.Tensor, use_boundary = False):
        mu = self.softplus_mu(parameters[:, 0])
        alpha = self.softplus_alpha(parameters[:, 1])
        logits = mu * alpha
        mean = 1 / alpha * torch.exp(logits)
        std = mean / torch.sigmoid(-logits)
        return mean, std

    def emit_sample(self, parameters: torch.Tensor, v_batch: torch.Tensor, use_boundary = False):
        mu = self.softplus_mu(parameters[:, :, 0])
        alpha = self.softplus_alpha(parameters[:, :, 1])
        distribution = torch.distributions.negative_binomial.NegativeBinomial(total_count=1/alpha, logits=alpha*mu)
        pred = distribution.sample()
        return pred

    def emit_sample_one_step(self, parameters: torch.Tensor, v_batch: torch.Tensor, use_boundary = False):
        mu = self.softplus_mu(parameters[:, 0])
        alpha = self.softplus_alpha(parameters[:, 1])
        distribution = torch.distributions.negative_binomial.NegativeBinomial(total_count=1/alpha, logits=alpha*mu)
        pred = distribution.sample()
        return pred

**Comment by kb007** [comment]
well thanks for the code but  , Can you kindly tell me
```
def loss_fn(self, parameters: torch.Tensor, v_batch: torch.Tensor, labels: torch.Tensor, missing_value = 0, use_boundary = False):
    mu = self.softplus_mu(parameters[:, :, 0])
    alpha = self.softplus_alpha(parameters[:, :, 1])
    zero_index = (labels != missing_value)
    distribution = torch.distributions.negative_binomial.NegativeBinomial(total_count=1/alpha[zero_index], logits=alpha[zero_index]*mu[zero_index])
    likelihood = distribution.log_prob(labels[zero_index])
    return -torch.mean(likelihood)
```

I didn't get  the ' parameters ' , labels  part  is it my training set ? or it is coming from params

**Comment by zhykoties** [comment]
"parameters" is basically the concatenation of the two parameters for negative binomial distributions, namely mu and alpha. They are what your model outputs. Labels are fed in as a separate parameter (see function signature).

**Comment by hubert98** [comment]
How are emit_mean* and emit_sample* functions integrated in the code? Could you give some hints how to fully switch to negative binomial?
